### PR TITLE
feat: add support for executing `.ll` files directly

### LINKED ISSUE #4686: Allow to compile LLVM IR via `lfortran`
So that the following:
```
lfortran a.f90
```
Can also be done using:
```
lfortran a.f90 --show-llvm > a.ll
lfortran a.ll
```
And it would be equivalent. A hackish way to do this is via https://github.com/lfortran/lfortran/pull/4685. We should implement this as a supported official feature. That will greatly aid with debugging bugs like #4635.

Note: this must compile using our LLVM, exactly as we do already, just load the LLVM IR from disk, instead of constructing it in memory from ASR. It is not enough to just pass the LLVM IR to Clang, as Clang uses a different LLVM version and setting, and for example it does not reproduce the bug #4635.

### PR DESCRIPTION
**Pull request by kmr-srbh**
resolves #4686

### CODE REVIEW

File: src/bin/lfortran.cpp [hunk 1]
```diff
@@ -893,81 +893,87 @@ int compile_to_object_file(const std::string &infile,
         LCompilers::PassManager& lpm)
 {
     std::string input = read_file(infile);
+    LCompilers::LLVMEvaluator e(compiler_options.target);
+    if (!endswith(infile, ".ll")) {
```
Comment (certik) [inline src/bin/lfortran.cpp:897]: This function is becoming complicated and also we should not be deciding here based on a file extension, rather the decision what to do should be done in the main program, and here we just do one task per function. So let's rename the original `compile_to_object_file` to `compile_src_to_object_file`, and add a new function `compile_llvm_to_object_file` which will do the task of this PR.

File: src/bin/lfortran.cpp [hunk 2]
```diff
@@ -2464,6 +2470,13 @@ int main_app(int argc, char *argv[]) {
         return link_executable({tmp_o}, outfile, runtime_library_dir,
                 backend, static_link, shared_link, link_with_gcc, true, arg_v, arg_L,
 		arg_l, linker_flags, compiler_options);
+    } else if (endswith(arg_file, ".ll")) {
+        err = compile_to_object_file(arg_file, tmp_o, false,
```
Comment (certik) [inline src/bin/lfortran.cpp:2474]: This will become:
```suggestion
        err = compile_llvm_to_object_file(arg_file, tmp_o, false,
```

**Comment by certik** [comment]
Awesome, thanks for working on this!

File: src/bin/lfortran.cpp (hunk 1 above)
Comment (kmr-srbh) [reply src/bin/lfortran.cpp:897]: Okay. :+1:

File: src/bin/lfortran.cpp [hunk 3]
```diff
@@ -925,7 +926,7 @@ int compile_to_object_file(const std::string &infile,
     }
 
     // ASR -> LLVM
-    LCompilers::LLVMEvaluator e(compiler_options.target);
+
```
Comment (certik) [inline src/bin/lfortran.cpp:929]: You can probably revert this change, it seems it's easier to understand if the `e` is declared closer to the block where it is used?

**Review by certik — APPROVED** [review]
I think this is nice and clean now, thanks!

Left a minor comment, you can address it either way and merge.

File: src/bin/lfortran.cpp (hunk 3 above)
Comment (kmr-srbh) [reply src/bin/lfortran.cpp:929]: Oh, this change is from the previous implementation I had which used file extensions. We required this to be declared outside the `if` block at the top.

It is not required now. I will revert it. :+1:

**Comment by certik** [comment]
Awesome. One other issue that we can fix in a subsequent PR is to report errors from LLVM if the input is incorrect. Right now I think it segfaults.

**Comment by kmr-srbh** [comment]
> Awesome. One other issue that we can fix in a subsequent PR is to report errors from LLVM if the input is incorrect. Right now I think it segfaults.

Yes, we can do that. :+1:
